cuda/src/memory/cuda/resources/terminal/device_resource.cpp

```diff
@@ -0,0 +1,35 @@
+#include <memory>
+#include <memory_resource>
+
+#include <cuda_runtime.h>
+
+#include "vecmem/memory/cuda/resources/terminal/device_resource.hpp"
+
+namespace vecmem::memory::resources::terminal {
+    void * cuda_device_resource::do_allocate(
+        std::size_t bytes,
+        std::size_t
+    ) {
+        void * res;
+        cudaMalloc(&res, bytes);
+        return res;
+    }
+
+    void cuda_device_resource::do_deallocate(
+        void * p,
+        std::size_t,
+        std::size_t
+    ) {
+        cudaFree(p);
```

Comment:
A couple of points:
  - I would call the class `vecmem::cuda::device_resource`. Let's not get too crazy with the namespaces. (The whole project is about memory management. I don't think we should add a `memory` sub-namespace at all.)
  - With that class name this file should be called `device_resource.cu`. So that CMake would recognise it as a CUDA file. (And also, pay attention to naming the files the same as the class that they hold!)
  - Please use `VECMEM_CUDA_ERROR_CHECK(...)` on all of the CUDA calls.

Reply:
Are you sure you want this to be compiled with the CUDA compiler? There is no device code here, so I don't personally see the need for it.

Reply:
It makes the CMake configuration way easier if we use the `.cu` file extension. As CMake then automatically makes the code see `<cuda_runtime.h>`, and link the code against the CUDA runtime library. Otherwise we need to do all of those by hand.

So while this is a slight overhead, I just got into the habit of putting all of my CUDA code (even the code that's only running on the host) into `.cu` files.

Also, I've set up the CMake configuration to do the same for `.hip` and `.sycl` files. There I also want to use the special extensions for any code that interacts with the runtime of those languages...